src/main/kotlin/org/rust/lang/utils/ProjectCache.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.utils
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.psi.util.CachedValue
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import com.intellij.util.containers.ContainerUtil
+import java.util.concurrent.ConcurrentMap
+
+class ProjectCache<in T, R>(cacheName: String) {
+    init {
+        if (!registered.add(cacheName)) {
+            error("""
+                ProjectCache `$cacheName` is already registered.
+                Make sure ProjectCache is static, that is, put it inside companion object.
+            """.trimIndent())
+        }
+    }
+
+    fun getOrPut(project: Project, key: T, defaultValue: () -> R): R {
+        val cache = CachedValuesManager.getManager(project)
+            .getCachedValue(project, cacheKey, provider, false)
+        return cache.getOrPut(key) { defaultValue() }
+    }
+
+    private val cacheKey: Key<CachedValue<ConcurrentMap<T, R>>> = Key.create(cacheName)
```

Comment:
Should we put fields before methods?

Reply:
Yeah, style guide says:

> Do not sort the method declarations alphabetically or by visibility

https://github.com/JetBrains/kotlin-web-site/blob/38a85989b11977559fad17b60001cd96736d9470/pages/docs/reference/coding-conventions.md#class-layout

will rebase, 👍

Reply:
Addressed!